### DISCUSSION TITLE
[Solved] More verbose error reporting

### DISCUSSION THREAD

**Discussion by Amenel** [Extended-Thunder/send-later#722] [Ideas]
Hi,
I was wondering if it would be possible to have a more verbose error reporting when an error occurs with Send Later. I'm asking this because I'm using a brand new computer, on which I have the latest Thunderbird (version 115.12.2) and Send Later (10.5.1 from two days ago) and the single email I have created to test Send Later never left the Drafts folder.

The email is from one of my own addresses to my Gmail address. I opened the error console and this is what I get there:

```
Error processing message 
Object { id: 1, date: Date Sat Jul 20 2024 09:25:27 GMT+0200 (heure d’été d’Europe centrale), author: "<amenel@…>", recipients: (1) […], ccList: [], bccList: [], subject: "Message planifié en sortie", read: true, new: false, headersOnly: false, … }
​
author: "<amenel@…>"
​
bccList: Array []
​
ccList: Array []
​
date: Date Sat Jul 20 2024 09:25:27 GMT+0200 (heure d’été d’Europe centrale)
​
external: false
​
flagged: false
​
folder: Object { accountId: "account1", name: "Drafts", path: "/Drafts", … }
​
headerMessageId: "406a4be2-611e-4e15-9c0e-887a03c71183@…"
​
headersOnly: false
​
id: 1
​
junk: false
​
junkScore: 0
​
new: false
​
read: true
​
recipients: Array [ "<…@gmail.com>" ]
​
size: 762
​
subject: "Message planifié en sortie"
​
tags: Array []
​
<prototype>: Object { … }
 Error: Error importing message: undefined
    messageImport moz-extension://f11d3733-401e-4f9a-a58d-2f6d5b9fbfee/utils/static.js:1799
    success moz-extension://f11d3733-401e-4f9a-a58d-2f6d5b9fbfee/background.js:841
    tbIsVersion moz-extension://f11d3733-401e-4f9a-a58d-2f6d5b9fbfee/utils/static.js:78
    tb115 moz-extension://f11d3733-401e-4f9a-a58d-2f6d5b9fbfee/utils/static.js:95
    doSendMessage moz-extension://f11d3733-401e-4f9a-a58d-2f6d5b9fbfee/background.js:838
    possiblySendMessage moz-extension://f11d3733-401e-4f9a-a58d-2f6d5b9fbfee/background.js:1105
    mainLoop moz-extension://f11d3733-401e-4f9a-a58d-2f6d5b9fbfee/background.js:2783
    pageResults moz-extension://f11d3733-401e-4f9a-a58d-2f6d5b9fbfee/utils/tools.js:321
    forAllDrafts moz-extension://f11d3733-401e-4f9a-a58d-2f6d5b9fbfee/utils/tools.js:320
    mainLoop moz-extension://f11d3733-401e-4f9a-a58d-2f6d5b9fbfee/background.js:2782
    alarmsListener moz-extension://f11d3733-401e-4f9a-a58d-2f6d5b9fbfee/background.js:2862
static.js:150:28
```     

It appears as though one of the variables is undefined. However, I haven't been able to trigger a breakpoint at the appropriate location in the source code because the Send Later code is not executed as long as the developer tools window is open. My intention was to determine what is "undefined" and proceed from there.

Is it possible to have a more detailed error reporting, so that one can know which actions to take? For instance, if there is a misconfiguration issue, the user my resolve the issue that is preventing Send Later from working the way it should.
Or else, is it actually possible to me to trigger a breakpoint in Send Later?

Thank you for any input!

**Comment by jikamens** [comment]
The `Error importing message: undefined` message you see there comes from Thunderbird, not from Send Later. I have no control over the fact that there is no useful explanation of what went wrong.

Generally speaking, the stuff that shows up in the error console is not intended to be user-friendly. When the add-on knows what to tell the user about a problem it puts it in a pop-up window; the stuff that gets logged is mostly for debugging, and I am not going to put a lot of effort into making it more user-friendly, because that's not the point.

You can't debug add-on code in the developer tools. You have to open the add-ons page, click the gear and select "Debug add-ons", click "Inspect" next to the add-on you want to debug, and then go to the Debugger tab of the inspection window that pops up.

I'm stumped about what's going wrong here, for the same reason you are: Thunderbird isn't providing any useful information about what's going wrong.

Is this problem reliably reproducible or intermittent? Have you been able to use Send Later before on other computers, or is this your first time trying to use it? What OS is running on your computer?

Note that the current version of Send Later is 128.0.1, not 115.12.2, though it appears they haven't pushed that out to the update channel for everyone yet. You could try TB128 and see if it still has problems. You could also try the current pre-release on GitHub, which has one bug fix that might help:

This release is not yet available for download from addons.thunderbird.net, but you can download and install it from its [GitHub release page](https://github.com/Extended-Thunder/send-later/releases/tag/v10.5.2) as follows:

* Download [send_later_10.5.2.xpi](https://github.com/Extended-Thunder/send-later/releases/download/v10.5.2/send_later_10.5.2.xpi) to your computer.
(If you are using Firefox, make sure to right click on the link and select "Save Link As..." rather than just clicking on it, because otherwise Firefox will try to install it as a Firefox add-on.)
* Open the Add-ons page in Thunderbird ("Add-ons and Themes" from the corner "hamburger" menu).
* Click the gear icon at the top and select "Install Add-on From File...".
* Browse to and select the downloaded file.
* Click through the installation pop-ups.
* You can delete the downloaded file once you've installed it into Thunderbird.

To subscribe to future beta releases (we love beta testers!), download and install [send_later_beta_10.5.2.xpi](https://github.com/Extended-Thunder/send-later/releases/download/v10.5.2/send_later_10.5.2_beta.xpi) instead. The advantage is that if there's a bug that affects your workflow you'll help find it and get it fixed quickly. The disadvantage is that bugs are a bit more likely in beta releases. You can unsubscribe from beta releases at any time by downloading and installing send_later.xpi or installing from [addons.thunderbird.net](https://addons.thunderbird.net/thunderbird/addon/send-later-3/).

Note that this is a prerelease. While we make every effort to ensure that prereleases are stable, they are a bit more likely to have bugs, so proceed with caution.

**Comment by Amenel** [comment]
Yes, I understand not investing time on the underlying system on top of which an extension is running.

The problem occurs consistently on this Windows 11 laptop: I haven't been able to schedule a single message with Send Later so far. I used to run Send Later many years ago on a previous laptop and since this is my only computer now, I think that my previous experience is not relevant.

I've successfully debugged 10.5.2 on the current Thunderbird (115.12.2) and it turns out that `localFolder.id` is undefined. Rightfully so, as this is the object being handled at line 1801 of static.js (i.e. `await SLStatic.tb128(localFolder.id, localFolder),`):

```
accountId: "account2"
name: "Outbox"
path: "/Unsent Messages"
subFolders: []
type: "outbox"
<prototype>: {…}
```

It has no `id` field. However, I could see among the variables that `localAccount` does have an `id` field.

I checked the filesystem for my "Local Folders" account. On the disk, it does have a folder "Unsent Messages" (which maps to "Outbox" in the Thunderbird UI) and a file "Unsent Messages.msf".

I'll be trying TB128 in a few days.

Thanks for your reaction, I appreciate it.

**Comment by jikamens** [comment]
`SLStatic.tb128(localFolder.id, localFolder)` is a compatibility shim. The first expression is only evaluated if the Thunderbird version is 128 or newer; for older versions the second expression is evaluated instead. This is not the root cause of the problem you are observing.

At first glance my guess is that you have too many files in your Windows temporary directory and you should use the disk cleanup utility to clean up temporary files. If that doesn't work we can try more troubleshooting.

**Comment by Amenel** [comment]
I cleaned up the temp files and folders, around 135 entries in %TEMP%, by running the disk cleanup utility. The error message remains the same.

I register that the first expression will be evaluated if TB is 128 or newer. I wonder then about when the JavaScript process encounters the code below in static.js. Won't it try to "collect" the arguments to the function call and fail with the error message "undefined" due to field `id` not being present in `localFolder`? By "collect", I mean "identify the variables that will potentially be evaluated later on", if that makes sense.

```
    let newMsgHeader = await messenger.messages.import(
      file,
      await SLStatic.tb128(localFolder.id, localFolder),
      properties,
    );
```

Also, while the breakpoint was triggered, I could see that the message.eml file was indeed created in the temp directory.

I'll try TB128 out in two days. If that fails too, I'll chalk this issue up to my specific configuration and reinstall Thunderbird. I'll report about all of that here.

Thanks for your work on this add-on!

**Comment by jikamens** [comment]
JavaScript is not a civilized programming language. It does not care whether object properties are defined or undefined. It is not an error to reference an undefined property; it's just undefined.

Compare  referencing an undefined property, which is not an error, vs. referencing an undefined variable, which is:

```
> try { console.log(o.no_such_property); } catch { console.log("An error was thrown"); }
undefined
undefined
> try { console.log(no_such_variable); } catch { console.log("An error was thrown"); }
An error was thrown
undefined
> 
```

The "undefined" message you're seeing in that error in the console isn't about the property being undefined. It's about the error message being undefined.

I don't know what's causing this problem, but it's not the undefined property on the folder object.

Separate from trying TB128, please try creating a new Thunderbird profile in TB115 and seeing if you can reproduce the issue in it. It is possible that this is a profile corruption issue.

**Comment by Amenel** [comment]
I've done a few tests. Here is the report.

I moved to TB128 before reading you :-( 

I've met the same issue. 

However, the issue **is not present** with a new profile. I've recreated the account with the email address touched by this issue, and used Send Later. No issues. I've then recreated the other email accounts and rescheduled a message with one of these other email accounts. No issue.
I then had each account in the new profile reuse the folder that was used in the old profile, starting with Local Folders, and the issue resurfaced. I concluded that there is a filesystem issue with the Local Folders directory of the old profile. By investigating the directory and making several tests, I've found the cause (and the fix).
Apparently, Thunderbird expects an "Unsent Messages" _file_ but there was a folder by that name. I guess that since two objects on the filesystem cannot have the same name in the same parent folder, the filesystem threw and error of some kind. The file was simple: I deleted the folder.

In the meantime, I have been able to identify the caught exception. It had id 2147500037 and no message, which you had already correctly inferred.

Again, thank you for your help.
